test/robustness/within/partition_within.cpp

```diff
@@ -311,27 +242,46 @@ void call_within(std::size_t size, std::size_t count)
         return elapsed;
     };
 
-    point_in_ring_visitor count_visitor;
-    {
 #if defined(TEST_WITH_SVG)
-
-        using partition_box_visitor_type = svg_visitor<box_type, std::vector<point_item>, std::vector<ring_item<ring_type>>>;
-        partition_box_visitor_type partition_box_visitor(size, points, rings);
-#else
-        using partition_box_visitor_type = bg::detail::partition::visit_no_policy;
-        partition_box_visitor_type partition_box_visitor;
+    std::size_t index = 0;
+    std::vector<box_type> boxes;
+    auto box_visitor = [&](auto const& box, int level)
+        {
+            create_svg(size, points, rings, index++, box, level, boxes);
+        };
+#else            
+    auto box_visitor = [&](auto const& , int ) {};
 #endif
 
+    std::size_t quadtree_count = 0;
+    {
         auto const start = std::chrono::steady_clock::now();
 
-        bg::partition
-            <
-                box_type,
-                bg::detail::partition::include_all_policy,
-                bg::detail::partition::include_all_policy
-            >::apply(points, rings, count_visitor, expand_for_point(), overlaps_point(),
-                    expand_for_ring(), overlaps_ring(), 16, partition_box_visitor);
-        report("Partition", start, count_visitor.count);
+        bg::experimental::quadtree_partition<box_type>(points, 
+            rings, 
+            [](auto& box, auto const& point) { bg::expand(box, point); },
+            [](auto const& box, auto const& point) { return ! bg::disjoint(point, box); },
+            [](auto& box, auto const& item) { bg::expand(box, item.box); },
+            [](auto const& box, auto const& item) 
+            {
+                typename bg::strategy::disjoint::services::default_strategy
+                    <
+                        box_type, box_type
+                    >::type strategy;
+                return ! bg::detail::disjoint::disjoint_box_box(box, item.box, strategy);
+            },
+            [&quadtree_count](auto const& p, auto const& ri)
+            {
+                if (bg::within(p, ri.ring))
+                {
+                    quadtree_count++;
+                }
+                return true;
+            },
+            box_visitor
+            );
```

Comment:
This shows how the function could be used.
5 lambdas (2 expand, 2 overlaps, 1 main) plus an optional lambda (called per box),
instead of the earlier policies with `apply`

Reply:
The program itself creates SVG's like this:

![image](https://user-images.githubusercontent.com/334849/236930381-1cdcceb8-979a-4f59-9f3f-150bd9a3189c.png)

Reply:
A lot better! I prefer function objects over objects with `apply()` function.

One nitpick. This is not quadtree, it's kdtree, a node in quadtree is divided into 4 children, here a division is done into 2, once per dimension. Furthermore partition is k-dimensional, quadtree is 2-dimensional.

Side note, I was thinking about redesigning the API for a long time and use features from the rtree. Note e.g. that in your example you shouldn't be forced to pass function objects defining how to expand a box or perform intersection. These operations are always done the same way. We probably also never use box visitors, besides debugging or visualization so box visitor could be hidden as well. The only additional information in your example that's needed is the way of getting a box from item (taking point from point should work by default). In rtree this is done by IndexableGetter. There are also additional things like the type of operation (in this case intersects) and strategies (for defining CS) but they could be optional as well. I'm not saying to do this, only sharing what I was thinking about.

Reply:
Thanks!

I think it's not exactly a kdtree either. You are right, we divide into halves and not into 4. But always at the middle, in kdtree that it is not necessarily the case. Anyway, I can rename it of course. But it cannot be called `partition`... (unless we replace it).
Plus we have those additional boxes for cases where geometries exceed the halves. But sure, they are not part of the quadtree either.

I see your point. Yes, taking the box from an object is usually enough. Or point. But it could be done in a different way as well. For example: a box with a margin (for (max) distance calculations). It's flexible now. But that flexibility is probably used rarely, if at all. We use partition only internally, I think. It was introduced before our index so at that moment it was not a choice. So yes, it can be a bit simpler. The box visitor is indeed only used for debugging (at least at this moment).

Anyway, additionally, if we go this way, we don't need those visitor-adaptor(s), we can deprecate the old implementation and let the new one immediately use function objects.

Reply:
Splitting planes can be placed at various locations in a kd-tree. Classic approach is to place them at the median object. Another strategy is to cut empty space out. There is also version using surface area heuristic. There are probably more. It depends on a use case. Splitting in half is probably a poor choice in general good only for objects distributed uniformly but such tree is still kd-tree.

Btw, in kd-tree splitting doesn't have to be done one dimension after another. The space can be split e.g. in X axis cutting out empty space and then again in X axis at the middle object, etc. But no matter what strategy is used this is still kd-tree.

Regarding quadtrees. There are versions of quadtrees have splitting planes at different locations than geometric middle of a node. AFAIR there are also quadtrees with child nodes partially overlapping each other which makes them similar to rtrees.

Obviously partition is neither quadtree nor kd-tree because it's not a data structure.

Reply:
It's used internally around 12 times.
Unfortunately it's not in namespace `detail` so officially it is part of the interface.